### DISCUSSION TITLE
[turborepo] with-docker: Discarding source files following build

### DISCUSSION THREAD

**Discussion by 5E7EN** [vercel/turborepo#7123] [Help] [answered]
### Which project is this feature idea for?

Turborepo

### Describe the feature you'd like to request

Upon reviewing the [with-docker](https://github.com/vercel/turbo/tree/main/examples/with-docker) example (in line with my own project), I have got a question regarding best practice for the final production image.

Here's a brief overview of my build process (in the form of a multi-stage Dockerfile):

Stage 1:
1. Copy all source files into image
2. Prune files to only include app and its dependencies (`turbo prune --scope=api --docker`)

Stage 2:
1. Copy pruned package.json files from stage 1 (from `/app/out/json`)
2. Install all dependencies (yarn install)
3. Copy rest of pruned source files from stage 2 (from `/app/out/full`)
4. Build project (`turbo build --filter=api`)

Stage 3:
1. Copy all files from stage 2 which includes the built artifacts and source files (from `/app`)
2. Run application

---

My question lies beyond step 1 in stage 3. I don't want to include the source files in the final production image.
The build output of step 4 in stage 2 produces a `dist` folder in the main application's directory, as well as in each of it's dependencies (remember, this is a monorepo).

What's the most efficient way to only end up with the built application files (and its built dependencies) - without having to manually copy the `dist` folders from each directory into the stage 3 image (since that requires explicitly defining each dependency the app depends on, which I want to avoid)? 

### Describe the solution you'd like

Perhaps the likes of a `turbo prune --production-source` flag that can be executed after the build to remove the original source `.ts` files, while keeping what's necessary to run the application (dist, package.json, etc).

### Describe alternatives you've considered

As part of the Dockerfile in the final image, deleting all matches of `*.ts` after copying the build + source from the previous image (following step 1 in stage 3).

**Comment by NicholasLYang** [comment]
Hi @5E7EN, thanks for the issue. I'll bring this up with the team. One temporary option you could explore is using the `--dry-run=json` output to traverse the graph and move the `dist` folders manually.

**Comment by 5E7EN** [comment]
@NicholasLYang thanks for the suggestion.
~~I can't seem to have `turborepo build --scope=api --dry-run=json` list my `common` package as a dependency (even though it is a dependency in the package.json of `api`)...~~

~~How does turborepo determine the dependencies when building?~~

Got this working by using ` turborepo build --scope=api... --dry-run=json` instead.

However, there are likely cases/packages that need other build artifacts other than just the dist/ folder (and package.json). How do you suggest I go about this?
Is it even considered bad practice to include the source files and other dev files in the production image?

There's gotta be some kind of industry standard solution that I'm missing here...

**Comment by NicholasLYang** [comment]
Could you expand on your use case here? Are you producing an application like a front-end app or a server? Or are you producing a bunch of libraries that are consumed later on? 

For the first option, you could use a bundler to combine all of the artifacts and ship a bundled application. For the second option, I think it's perfectly fine to ship the source files in the production code.

**Comment by lacherogwu** [comment]
My workaround is: 
I store all my source code in the `src/` directory and include a step in the Dockerfile to remove all `src/` directories.

```Dockerfile
...

# after build step
RUN find . -type d -name 'node_modules' -prune -o -type d -name 'src' -print | xargs rm -rf

...
```

**Comment by mehulkar** [comment, marked as answer]
I've converted this to a discussion, because this is outside the scope of the existing feature set. Turbo lets you prune the packages in your monorepo to get a subset of your source files, but it doesn't know or care about how your `build` scripts (or any tasks) work, so it cannot make any decisions about that. Your solution to add a script to remove source files after building is the best one!
